File: bzcat/Bzcat.java
//van github/volueinsight/bzip2

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.InputStream;
import java.io.IOException;
import java.io.OutputStream;

public class Bzcat
{
    class BZip2BitInputStream
    {
        private final InputStream inputStream;
        private int bitBuffer;
        private int bitCount;

        public boolean readBoolean() throws IOException
        {
            int bitBuffer = this.bitBuffer;
            int bitCount = this.bitCount;

            if (bitCount > 0) {
                bitCount--;
            }
            else
            {
                int byteRead = this.inputStream.read();
                
                if (byteRead < 0) {
                    throw new BZip2Exception ("Insufficient data");
                }

                bitBuffer = (bitBuffer << 8) | byteRead;
                bitCount += 7;
                this.bitBuffer = bitBuffer;
            }

            this.bitCount = bitCount;
            return ((bitBuffer & (1 << bitCount))) != 0;
        }

        public int readUnary() throws IOException
        {
            int bitBuffer = this.bitBuffer;
            int bitCount = this.bitCount;
            int unaryCount = 0;
    
            for (;;) {
                if (bitCount > 0) {
                    bitCount--;
                } else  {
                    int byteRead = this.inputStream.read();
    
                    if (byteRead < 0) {
                        throw new BZip2Exception ("Insufficient data");
                    }
    
                    bitBuffer = (bitBuffer << 8) | byteRead;
                    bitCount += 7;
                }
    
                if (((bitBuffer & (1 << bitCount))) == 0) {
                    this.bitBuffer = bitBuffer;
                    this.bitCount = bitCount;
                    return unaryCount;
                }
                unaryCount++;
            }
        }

        public int readBits (final int count) throws IOException
        {
            int bitBuffer = this.bitBuffer;
            int bitCount = this.bitCount;
    
            if (bitCount < count) {
                while (bitCount < count) {
                    int byteRead = this.inputStream.read();
    
                    if (byteRead < 0) {
                        throw new BZip2Exception ("Insufficient data");
                    }
    
                    bitBuffer = (bitBuffer << 8) | byteRead;
                    bitCount += 8;
                }
    
                this.bitBuffer = bitBuffer;
            }
    
            bitCount -= count;
            this.bitCount = bitCount;
            return (bitBuffer >>> bitCount) & ((1 << count) - 1);
        }

        public int readInteger() throws IOException {
            return (readBits (16) << 16) | (readBits (16));
        }

        public BZip2BitInputStream (final InputStream inputStream)
        {
            this.inputStream = inputStream;
        }
    }

    interface BZip2Constants
    {
        static final int HUFFMAN_ENCODE_MAXIMUM_CODE_LENGTH = 20;
        static final int HUFFMAN_DECODE_MAXIMUM_CODE_LENGTH = 23;
        static final int HUFFMAN_SYMBOL_RUNA = 0;
        static final int HUFFMAN_SYMBOL_RUNB = 1;
    }

    public class BZip2HuffmanStageDecoder
    {
        final BZip2BitInputStream _bis;
        final byte[] selectors;
        final int[] minimumLengths = new int[6];
        final int[][] _codeBases = new int[6][25];
        final int[][] codeLimits = new int[6][24];
        final int[][] _codeSymbols = new int[6][258];
        int currentTable;
        int groupIndex = -1;
        int groupPosition = -1;

        public int nextSymbol() throws IOException
        {
            // Move to next group selector if required
            if (((++this.groupPosition % 50) == 0))
            {
                this.groupIndex++;
                if (this.groupIndex == this.selectors.length)
                    throw new BZip2Exception ("Error decoding BZip2 block");
                
                this.currentTable = this.selectors[this.groupIndex] & 0xff;
            }

            final int currentTable = this.currentTable;
            final int[] tableLimits = this.codeLimits[currentTable];
            int codeLength = this.minimumLengths[currentTable];

            int codeBits = _bis.readBits(codeLength);
            for (; codeLength <= BZip2Constants.HUFFMAN_DECODE_MAXIMUM_CODE_LENGTH; codeLength++)
            {
                if (codeBits <= tableLimits[codeLength])
                {
                    return _codeSymbols[currentTable][
                            codeBits - _codeBases[currentTable][codeLength]];
                }
                codeBits = (codeBits << 1) | _bis.readBits (1);
            }

            // A valid code was not recognised
            throw new BZip2Exception ("Error decoding BZip2 block");

        }

        void _createHuffmanDecodingTables(final int alphabetSize,
            final byte[][] tableCodeLengths)
        {
            for (int table = 0; table < tableCodeLengths.length; table++)
            {
                final int[] tableBases = _codeBases[table];
                final int[] tableLimits = this.codeLimits[table];
                final int[] tableSymbols = _codeSymbols[table];

                final byte[] codeLengths = tableCodeLengths[table];
                int minimumLength = BZip2Constants.HUFFMAN_DECODE_MAXIMUM_CODE_LENGTH;
                int maximumLength = 0;

                // Find the minimum and maximum code length for the table
                for (int i = 0; i < alphabetSize; i++) {
                    maximumLength = Math.max (codeLengths[i], maximumLength);
                    minimumLength = Math.min (codeLengths[i], minimumLength);
                }

                this.minimumLengths[table] = minimumLength;

                // Calculate the first output symbol for each code length
                for (int i = 0; i < alphabetSize; i++)
                    tableBases[codeLengths[i] + 1]++;
                
                for (int i = 1; i < BZip2Constants.HUFFMAN_DECODE_MAXIMUM_CODE_LENGTH + 2; i++)
                    tableBases[i] += tableBases[i - 1];

                // Calculate the first and last Huffman code for each code length (codes at a given
                // length are sequential in value)
                int code = 0;
                for (int i = minimumLength; i <= maximumLength; i++)
                {
                    int base = code;
                    code += tableBases[i + 1] - tableBases[i];
                    tableBases[i] = base - tableBases[i];
                    tableLimits[i] = code - 1;
                    code <<= 1;
                }
                // Populate the mapping from canonical code index to output symbol
                int codeIndex = 0;
                for (int bitLength = minimumLength; bitLength <= maximumLength; bitLength++)
                    for (int symbol = 0; symbol < alphabetSize; symbol++)
                        if (codeLengths[symbol] == bitLength)
                            tableSymbols[codeIndex++] = symbol;

            }
        }



        public BZip2HuffmanStageDecoder (final BZip2BitInputStream bitInputStream,
            final int alphabetSize, final byte[][] tableCodeLengths, final byte[] selectors)
        {
            _bis = bitInputStream;
            this.selectors = selectors;
            this.currentTable = this.selectors[0];
            _createHuffmanDecodingTables (alphabetSize, tableCodeLengths);
        }
    }

    public final class CRC32
    {
        private int _table[] = new int[256];
        private int _crc = 0xffffffff;
        public int getCRC() { return ~_crc; }

        CRC32()
        {
            //https://github.com/gcc-mirror/gcc/blob/master/libiberty/crc32.c
            for (int i = 0, j, c; i < 256; i++)
            {
                for (c = i << 24, j = 8; j > 0; --j)
                    c = (c & 0x80000000) != 0 ? (c << 1) ^ 0x04c11db7 : (c << 1);
                _table[i] = c;
            }
        }

        public void updateCRC(final int value)
        {
            _crc = (_crc << 8) ^ _table[((_crc >> 24) ^ value) & 0xff];
        }
    }

    public class BZip2Exception extends IOException
    {
        private static final long serialVersionUID = -8931219115669559570L;
        public BZip2Exception (String reason) {
            super (reason);
        }
    }

    public class MoveToFront
    {
        private byte[] _mtf = new byte[256];

        public MoveToFront()
        {
            for (int i = 0; i < 256; ++i)
                _mtf[i] = (byte)i;
        }

        public byte indexToFront (final int index)
        {
            final byte value = _mtf[index];
            System.arraycopy (_mtf, 0, _mtf, 1, index);
            _mtf[0] = value;
            return value;
        }
    }

    public class BZip2BlockDecompressor
    {
        final BZip2BitInputStream bitInputStream;
        final CRC32 _crc = new CRC32();
        final int _blockCRC;
        int huffmanEndOfBlockSymbol;
        final byte[] huffmanSymbolMap = new byte[256];
        final int[] bwtByteCounts = new int[256];
        byte[] bwtBlock;
        int[] bwtMergedPointers;
        int bwtCurrentMergedPointer;
        int bwtBlockLength;
        int bwtBytesDecoded;
        int rleLastDecodedByte = -1;
        int rleAccumulator;
        int _rleRepeat;

        private BZip2HuffmanStageDecoder readHuffmanTables() throws IOException
        {
            final byte[] huffmanSymbolMap = this.huffmanSymbolMap;
            final byte[][] tableCodeLengths = new byte[6][258];
            int huffmanUsedRanges = bitInputStream.readBits(16);
            int huffmanSymbolCount = 0;

            for (int i = 0; i < 16; i++)
                if ((huffmanUsedRanges & ((1 << 15) >>> i)) != 0)
                    for (int j = 0, k = i << 4; j < 16; j++, k++)
                        if (bitInputStream.readBoolean())
                            huffmanSymbolMap[huffmanSymbolCount++] = (byte)k;
            
            int endOfBlockSymbol = huffmanSymbolCount + 1;
            this.huffmanEndOfBlockSymbol = endOfBlockSymbol;

            /* Read total number of tables and selectors*/
            final int totalTables = bitInputStream.readBits (3);
            final int totalSelectors = bitInputStream.readBits (15);

            /* Read and decode MTFed Huffman selector list */
            final MoveToFront tableMTF = new MoveToFront();
            final byte[] selectors = new byte[totalSelectors];
            for (int selector = 0; selector < totalSelectors; selector++)
                selectors[selector] = tableMTF.indexToFront (bitInputStream.readUnary());

            /* Read the Canonical Huffman code lengths for each table */
            for (int table = 0; table < totalTables; table++)
            {
                int currentLength = bitInputStream.readBits (5);
                for (int i = 0; i <= endOfBlockSymbol; i++)
                {
                    while (bitInputStream.readBoolean())
                        currentLength += bitInputStream.readBoolean() ? -1 : 1;
                    tableCodeLengths[table][i] = (byte)currentLength;
                }
            }

            return new BZip2HuffmanStageDecoder(bitInputStream, endOfBlockSymbol + 1,
                                                tableCodeLengths, selectors);
        }

        private void decodeHuffmanData (final BZip2HuffmanStageDecoder huffmanDecoder)
            throws IOException
        {
            final byte[] bwtBlock = this.bwtBlock;
            final byte[] huffmanSymbolMap = this.huffmanSymbolMap;
            final int streamBlockSize = this.bwtBlock.length;
            final int huffmanEndOfBlockSymbol = this.huffmanEndOfBlockSymbol;
            final int[] bwtByteCounts = this.bwtByteCounts;
            final MoveToFront symbolMTF = new MoveToFront();
            int bwtBlockLength = 0;
            int repeatCount = 0;
            int repeatIncrement = 1;
            int mtfValue = 0;

            for (;;)
            {
                final int nextSymbol = huffmanDecoder.nextSymbol();

                if (nextSymbol == BZip2Constants.HUFFMAN_SYMBOL_RUNA)
                {
                    repeatCount += repeatIncrement;
                    repeatIncrement <<= 1;
                    continue;
                }
                if (nextSymbol == BZip2Constants.HUFFMAN_SYMBOL_RUNB)
                {
                    repeatCount += repeatIncrement << 1;
                    repeatIncrement <<= 1;
                    continue;
                }
                
                if (repeatCount > 0)
                {
                    if (bwtBlockLength + repeatCount > streamBlockSize)
                        throw new BZip2Exception ("BZip2 block exceeds declared block size");
                
                    final byte nextByte = huffmanSymbolMap[mtfValue];
                    bwtByteCounts[nextByte & 0xff] += repeatCount;
                    while (--repeatCount >= 0)
                        bwtBlock[bwtBlockLength++] = nextByte;

                    repeatCount = 0;
                    repeatIncrement = 1;
                }

                if (nextSymbol == huffmanEndOfBlockSymbol)
                    break;

                if (bwtBlockLength >= streamBlockSize)
                    throw new BZip2Exception ("BZip2 block exceeds declared block size");

                mtfValue = symbolMTF.indexToFront (nextSymbol - 1) & 0xff;
                final byte nextByte = huffmanSymbolMap[mtfValue];
                bwtByteCounts[nextByte & 0xff]++;
                bwtBlock[bwtBlockLength++] = nextByte;
            
            }

            this.bwtBlockLength = bwtBlockLength;
        }

        private int decodeNextBWTByte()
        {
            int mergedPointer = this.bwtCurrentMergedPointer;
            int nextDecodedByte =  mergedPointer & 0xff;
            this.bwtCurrentMergedPointer = this.bwtMergedPointers[mergedPointer >>> 8];
            this.bwtBytesDecoded++;
            return nextDecodedByte;
        }

        public int read()
        {
            while (_rleRepeat < 1)
            {
                if (this.bwtBytesDecoded == this.bwtBlockLength)
                    return -1;

                int nextByte = decodeNextBWTByte();

                if (nextByte != this.rleLastDecodedByte)
                {
                    // New byte, restart accumulation
                    this.rleLastDecodedByte = nextByte;
                    _rleRepeat = 1;
                    this.rleAccumulator = 1;
                    _crc.updateCRC (nextByte);
                }
                else if (++this.rleAccumulator == 4)
                {
                    // Accumulation complete, start repetition
                    _rleRepeat = decodeNextBWTByte() + 1;
                    this.rleAccumulator = 0;

                    for (int i = 0; i < _rleRepeat; ++i)
                        _crc.updateCRC(nextByte);
                    //_crc.updateCRC (nextByte, _rleRepeat);
                }
                else
                {
                    _rleRepeat = 1;
                    _crc.updateCRC(nextByte);
                }
            }

            _rleRepeat--;
            return this.rleLastDecodedByte;
        }

        public int checkCRC() throws IOException
        {
            if (_blockCRC != _crc.getCRC())
                throw new BZip2Exception ("BZip2 block CRC error");

            return _crc.getCRC();
        }

        public BZip2BlockDecompressor(final BZip2BitInputStream bitInputStream,
                            final int blockSize) throws IOException
        {
            this.bitInputStream = bitInputStream;
            this.bwtBlock = new byte[blockSize];

            final int bwtStartPointer;

            // Read block header
            _blockCRC = this.bitInputStream.readInteger();
            boolean blockRandomised = this.bitInputStream.readBoolean();

            if (blockRandomised)
                throw new BZip2Exception("Randomised blocks not supported.");

            bwtStartPointer = this.bitInputStream.readBits (24);

            // Read block data and decode through to the Inverse Burrows Wheeler Transform stage
            BZip2HuffmanStageDecoder huffmanDecoder = readHuffmanTables();
            decodeHuffmanData (huffmanDecoder);
            //initialiseInverseBWT (bwtStartPointer);
            final byte[] bwtBlock  = this.bwtBlock;
            final int[] bwtMergedPointers = new int[this.bwtBlockLength];
            final int[] characterBase = new int[256];

            if ((bwtStartPointer < 0) || (bwtStartPointer >= this.bwtBlockLength))
                throw new BZip2Exception ("BZip2 start pointer invalid");

            // Cumulatise character counts
            System.arraycopy (this.bwtByteCounts, 0, characterBase, 1, 255);
            for (int i = 2; i <= 255; i++)
                characterBase[i] += characterBase[i - 1];

            for (int i = 0; i < this.bwtBlockLength; i++)
            {
                int value = bwtBlock[i] & 0xff;
                bwtMergedPointers[characterBase[value]++] = (i << 8) + value;
            }

            this.bwtBlock = null;
            this.bwtMergedPointers = bwtMergedPointers;
            this.bwtCurrentMergedPointer = bwtMergedPointers[bwtStartPointer];
        }
    }

    void decompress(InputStream is, OutputStream os) throws IOException
    {
        BZip2BitInputStream bis = new BZip2BitInputStream(is);

        int smarker1 = bis.readBits(16);
        int smarker2 = bis.readBits(8);
        int blockSize = (bis.readBits(8) - '0');
        int streamBlockSize = blockSize * 100000;
        int streamCRC = 0;

        while (true)
        {
            int marker1 = bis.readBits(24);
            int marker2 = bis.readBits(24);
 
            if (marker1 == 0x314159 && marker2 == 0x265359)
            {
                BZip2BlockDecompressor block = new BZip2BlockDecompressor(bis, streamBlockSize);
                
                for (int c; (c = block.read()) != -1;)
                    os.write(c);

                int blockCRC = block.checkCRC();
                streamCRC = ((streamCRC << 1) | (streamCRC >>> 31)) ^ blockCRC;
                continue;
            }
 
            if (marker1 == 0x177245 && marker2 == 0x385090)
            {
                os.flush();
                int crc = bis.readInteger();
                System.err.println(String.format("0x%08X 0x%08X", streamCRC, crc));
 
                if (crc != streamCRC)
                    throw new BZip2Exception ("BZip2 stream CRC error");
                 
                break;
            }
 
            throw new BZip2Exception ("BZip2 stream format error");
        }
    }

    void submain(String[] args) throws IOException
    {
        InputStream is = System.in;
        File inputFile;

        if (args.length == 1)
        {
            inputFile = new File(args[0]);
            if (!inputFile.exists() || !inputFile.canRead() || !args[0].endsWith(".bz2"))
            {
                System.err.println ("Cannot read file " + inputFile.getPath());
                System.exit (1);
            }
            is = new BufferedInputStream(new FileInputStream(inputFile));
        }
        decompress(is, System.out);
    }

    public static void main (String[] args) throws IOException
    {
        new Bzcat().submain(args);
    }
}

File: bzcat/bzcat.cpp
//This is a comment
//I love comments

#define FAST

#include <cstdint>
#include <cassert>
#ifdef FAST
#include <unistd.h>
#include <fcntl.h>
#else
#include <iostream>
#include <fstream>
#endif

#ifdef FAST
class istream
{
private:
    uint32_t _cap;
    uint32_t _head = 0, _tail = 0;
    uint8_t *_buf;
protected:
    int _fd;
public:
    ~istream() { delete[] _buf; }

    istream(int fd = -1, uint32_t capacity = 8192)
      : _cap(capacity), _buf(new uint8_t[capacity]), _fd(fd) { }

    int get()
    {
        if (_tail == _head)
        {
            ssize_t n = ::read(_fd, _buf, _cap);
            if (n < 1) return -1;
            _head = n;
            _tail = 0;
        }
        return _buf[_tail++];
    }
};

class ifstream : public istream
{
public:
    void close() { ::close(_fd); }
    void open(const char *fn) { _fd = ::open(fn, O_RDONLY); }
};

class ostream
{
public:
    virtual void put(char) { }
    virtual void flush() { }
    virtual ostream& operator<<(const char *) { return *this; }
};

class outstream : public ostream
{
    int _fd;
    uint32_t _cap;
    uint32_t _pos = 0;
    char *_buf;
public:
    ~outstream() { delete[] _buf; }
    void put(char c) override { if (_pos > _cap) flush(); _buf[_pos++] = c; }
    ostream& operator<<(const char *s) override { while (*s) put(*s++); return *this; }
    void flush() { ::write(_fd, _buf, _pos), _pos = 0; }

    outstream(int fd, uint32_t capacity = 8192)
      : _fd(fd), _cap(capacity), _buf(new char[capacity]) { }
};

class NullStream : public ostream
{
public:
    ostream& operator<<(const char *) override { return *this; }
};

static istream cin(0, 8192);
static outstream cout(1, 8192);
static outstream cerr(2, 8192);
static NullStream nullstream;
#else
class nullbuf : public std::streambuf
{
public:
    int overflow(int c) override { return c; }
};

using std::istream;
using std::ostream;
using std::ifstream;
using std::cin;
using std::cout;
using std::cerr;
static nullbuf nb;
static std::ostream nullstream(&nb);
#endif

class Toolbox
{
public:
    template <class T> static T min(T a, T b) { return b < a ? b : a; }
    template <class T> static T max(T a, T b) { return a < b ? b : a; }

    static char nibble(uint8_t n)
    { return n <= 9 ? '0' + char(n) : 'a' + char(n - 10); }

    static void hex32(unsigned dw, ostream &os)
    { for (unsigned i = 0; i <= 28; i += 4) os.put(nibble(dw >> 28 - i & 0xf)); }
};

class BitInputStream
{
    istream &_is;
    unsigned _window = 0, _bitCount = 0;
public:
    BitInputStream(istream &is) : _is(is) { }

    unsigned readBits24(unsigned n)
    {
        assert(n <= 24);
        while (_bitCount < n)
            _window = _window << 8 | _is.get(), _bitCount += 8;
        _bitCount -= n;
        return _window >> _bitCount & (1 << n) - 1;
    }

    unsigned readBits32(unsigned n)
    {
        assert(n <= 32);
        unsigned ret = 0;
        while (n)
        {
            unsigned foo = Toolbox::min(16U, n);
            ret = ret << foo | readBits24(foo), n -= foo;
        }
        return ret;
    }
};

class MoveToFront
{
    uint8_t _buf[256];
public:
    MoveToFront()
    { for (unsigned i = 0; i < 256; ++i) _buf[i] = i; }

    uint8_t indexToFront(unsigned i)
    { uint8_t val = _buf[i]; for (; i; --i) _buf[i] = _buf[i - 1]; return _buf[0] = val; }
};

class CRC32
{
    uint32_t _table[256];
    uint32_t _crc = 0xffffffff;
public:
    void update(uint8_t c) { _crc = _crc << 8 ^ _table[(_crc >> 24 ^ c) & 0xff]; }
    uint32_t crc() const { return ~_crc; }

    CRC32()
    {
        //https://github.com/gcc-mirror/gcc/blob/master/libiberty/crc32.c
        for (uint32_t i = 0, j, c; i < 256; ++i)
        {
            for (c = i << 24, j = 8; j > 0; --j)
                c = c & 0x80000000 ? c << 1 ^ 0x04c11db7 : c << 1;
            _table[i] = c;
        }
    }
};

class Table
{
    uint8_t _codeLengths[258];
    unsigned _pos = 0;
    unsigned _bases[25] = {0};
    unsigned _limits[24] = {0};
    unsigned _symbols[258] = {0};
    uint8_t _minLen = 23;
    uint8_t _maxLen = 0;
public:
    void read(BitInputStream &bis, unsigned symbolCount);
    uint8_t minLength() const { return _minLen; }
    uint32_t limit(uint8_t i) const { return _limits[i]; }
    uint32_t symbol(uint16_t i) const { return _symbols[i]; }
    uint32_t base(uint8_t i) const { return _bases[i]; }
};

//read the canonical Huffman code lengths for table
void Table::read(BitInputStream &bis, unsigned symbolCount)
{
    for (unsigned i = 0, c = bis.readBits24(5); i <= symbolCount + 1; ++i)
    {
        while (bis.readBits24(1))
            c += bis.readBits24(1) ? -1 : 1;
        _codeLengths[_pos++] = c;
    }

    for (unsigned i = 0; i < symbolCount + 2; ++i)
        _bases[_codeLengths[i] + 1]++;

    for (unsigned i = 1; i < 25; ++i)
        _bases[i] += _bases[i - 1];

    for (unsigned i = 0; i < symbolCount + 2; ++i)
    {
        _minLen = Toolbox::min(_codeLengths[i], _minLen);
        _maxLen = Toolbox::max(_codeLengths[i], _maxLen);
    }

    for (unsigned i = _minLen, code = 0; i <= _maxLen; ++i)
    {
        unsigned base = code;
        code += _bases[i + 1] - _bases[i];
        _bases[i] = base - _bases[i];
        _limits[i] = code - 1;
        code <<= 1;
    }

    for (unsigned i = 0, minLen = _minLen; minLen <= _maxLen; ++minLen)
        for (unsigned symbol = 0; symbol < symbolCount + 2; ++symbol)
            if (_codeLengths[symbol] == minLen)
                _symbols[i++] = symbol;
}

class Tables
{
    Table _tables[6];
    uint8_t *_selectors;
    uint32_t grpIdx, grpPos, curTbl;
public:
    ~Tables() { delete[] _selectors; }
    void read(BitInputStream &bis, uint32_t symbolCount);
    uint32_t nextSymbol(BitInputStream &bis);
};

void Tables::read(BitInputStream &bis, uint32_t symbolCount)
{
    uint8_t nTables = bis.readBits24(3);
    uint16_t nSelectors = bis.readBits24(15);
    _selectors = new uint8_t[nSelectors];
    MoveToFront tableMTF;
    
    for (uint32_t i = 0; i < nSelectors; ++i)
    {
        uint8_t u = 0;

        while (bis.readBits24(1))
            ++u;

        _selectors[i] = tableMTF.indexToFront(u);
    }

    for (uint32_t t = 0; t < nTables; ++t)
        _tables[t].read(bis, symbolCount);

    curTbl = _selectors[0], grpIdx = 0, grpPos = 0;
}

uint32_t Tables::nextSymbol(BitInputStream &bis)
{
    if (grpPos++ % 50 == 0)
        curTbl = _selectors[grpIdx++];

    unsigned i = _tables[curTbl].minLength();
    unsigned codeBits = bis.readBits24(i);

    for (;i <= 23; ++i)
    {
        if (codeBits <= _tables[curTbl].limit(i))
            return _tables[curTbl].symbol(codeBits - _tables[curTbl].base(i));

        codeBits = codeBits << 1 | bis.readBits24(1);
    }

    return 0;
}

class Block
{
    CRC32 _crc;
    uint32_t _dec = 0, _curp = 0, *_merged;
    uint8_t _nextByte();
public:
    uint32_t process(BitInputStream &bi, uint32_t blockSize, ostream &os);
};

uint8_t Block::_nextByte()
{       
    uint8_t ret = _curp & 0xff;
    _curp = _merged[_curp >> 8];
    ++_dec;
    return ret;
}

uint32_t Block::process(BitInputStream &bi, uint32_t blockSize, ostream &os)
{
    uint32_t _blockCRC = bi.readBits32(32);
    assert(bi.readBits24(1) == 0);
    unsigned bwtStartPointer = bi.readBits24(24), symbolCount = 0;
    unsigned bwtByteCounts[256] = {0};
    uint8_t symbolMap[256] = {0};

    for (unsigned i = 0, ranges = bi.readBits24(16); i < 16; ++i)
        if ((ranges & 1 << 15 >> i) != 0)
            for (unsigned j = 0, k = i << 4; j < 16; ++j, ++k)
                if (bi.readBits24(1))
                    symbolMap[symbolCount++] = uint8_t(k);

    Tables tables;
    tables.read(bi, symbolCount);
    uint8_t bwtBlock[blockSize], mtfValue = 0;
    MoveToFront symbolMTF;
    uint32_t _length = 0;

    for (unsigned n = 0, inc = 1;;)
    {
        unsigned nextSymbol = tables.nextSymbol(bi);

        if (nextSymbol == 0)
        {
            n += inc;
            inc <<= 1;
            continue;
        }

        if (nextSymbol == 1)
        {
            n += inc << 1;
            inc <<= 1;
            continue;
        }

        if (n > 0)
        {
            uint8_t nextByte = symbolMap[mtfValue];
            bwtByteCounts[nextByte] += n;
            ++n, inc = 1;

            while (--n >= 1)
                bwtBlock[_length++] = nextByte;
        }

        //end of block
        if (nextSymbol == symbolCount + 1)
            break;

        mtfValue = symbolMTF.indexToFront(nextSymbol - 1);
        uint8_t nextByte = symbolMap[mtfValue];
        bwtByteCounts[nextByte]++;
        bwtBlock[_length++] = nextByte;
    }

    _merged = new uint32_t[_length];
    unsigned characterBase[256] = {0};

    for (unsigned i = 0; i < 255; ++i)
        characterBase[i + 1] = bwtByteCounts[i];

    for (unsigned i = 2; i <= 255; ++i)
        characterBase[i] += characterBase[i - 1];

    for (unsigned i = 0; i < _length; ++i)
    {
        unsigned val = bwtBlock[i] & 0xff;
        _merged[characterBase[val]++] = (i << 8) | val;
    }

    _curp = _merged[bwtStartPointer];
    unsigned repeat = 0, acc = 0;
    int last = -1;

    while (true)
    {
        if (repeat < 1)
        {
            if (_dec == _length)
                break;

            uint8_t nextByte = _nextByte();

            if (nextByte != last)
            {
                last = nextByte, repeat = 1, acc = 1;
                _crc.update(nextByte);
            }
            else if (++acc == 4)
            {
                repeat = _nextByte() + 1, acc = 0;

                for (unsigned i = 0; i < repeat; ++i)
                    _crc.update(nextByte);
            }
            else
            {
                repeat = 1;
                _crc.update(nextByte);
            }
        }

        --repeat;
        os.put(last);
    }

    os.flush();
    delete[] _merged;
    assert(_blockCRC == _crc.crc());
    return _crc.crc();
}

int main(int argc, char **argv)
{
    static constexpr bool quiet = true;
    istream *is = &cin;
    ostream *os = &cout;
    ostream *msg = &cerr;
    ifstream ifs;

    if (quiet)
        msg = &nullstream;

    if (argc == 2)
        ifs.open(argv[1]), is = &ifs;

    BitInputStream bi(*is);
    assert(bi.readBits24(16) == 0x425a);
    bi.readBits24(8);
    uint8_t blockSize = bi.readBits24(8) - '0';
    uint32_t streamCRC = 0;

    while (true)
    {
        uint32_t marker1 = bi.readBits24(24), marker2 = bi.readBits24(24);

        if (marker1 == 0x314159 && marker2 == 0x265359)
        {
            Block b;
            uint32_t blockCRC = b.process(bi, blockSize * 100000, *os);
            streamCRC = (streamCRC << 1 | streamCRC >> 31) ^ blockCRC;
            continue;
        }

        if (marker1 == 0x177245 && marker2 == 0x385090)
        {
            uint32_t crc = bi.readBits32(32);
            assert(crc == streamCRC);
            *msg << "0x";
            Toolbox::hex32(crc, *msg);
            *msg << " 0x";
            Toolbox::hex32(streamCRC, *msg);
            *msg << "\r\n";
            msg->flush();
            break;
        }

        assert(false);
    }
    ifs.close();
    return 0;
}

